v2/example_test.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019, Google Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package gax_test
+
+import (
+	"context"
+	"time"
+
+	"github.com/googleapis/gax-go/v2"
+	"google.golang.org/grpc/codes"
+)
+
+const someRPCTimeout = 5 * time.Minute
+
+// Some result that the client might return.
+type fakeResponse struct{}
+
+// Some client that can perform RPCs.
+type fakeClient struct{}
+
+// PerformSomeRPC is a fake RPC that a client might perform.
+func (c *fakeClient) PerformSomeRPC(ctx context.Context) (*fakeResponse, error) {
+	// An actual client would return something meaningful here.
+	return nil, nil
+}
+
+func ExampleOnCodes() {
+	ctx := context.Background()
+	c := &fakeClient{}
+
+	// INTERNAL and UNKNOWN are typically safe to retry for idempotent RPCs.
+	retryer := gax.OnCodes([]codes.Code{codes.Unknown, codes.Unavailable}, gax.Backoff{
+		Initial:    time.Second,
+		Max:        32 * time.Second,
+		Multiplier: 2,
+	})
+
+	// It's recommended to set deadlines on RPCs and around retrying.
+	ctxWithTimeout, cancel := context.WithDeadline(ctx, time.Now().Add(someRPCTimeout))
+	defer cancel()
+
+	var err error
+	var resp *fakeResponse
+	for {
+		resp, err = c.PerformSomeRPC(ctxWithTimeout)
+		if err != nil {
```

Comment:
I'm a little worried by the nesting :(
My personal taste is to refactor into if-break and if-continue, but I'll defer to you and @tbpg.

Reply:
Suggested change, untested, including the return comment below. I might be missing something here.

It would be nice if `Retry` had a different API that would abstract some of this away (maybe it would take a `context.Context` and a `func() error` then handles the for loop)? Not sure how to get function return values.

```go
	var resp *fakeResponse
	for {
		var err error
		resp, err = c.PerformSomeRPC(ctx)
		if err == nil {
			break
		}
		delay, shouldRetry := retryer.Retry(err)
		if !shouldRetry {
			return fmt.Errorf("TODO")
		}
		if err = gax.Sleep(ctxWithTimeout, delay); err != nil {
			return fmt.Errorf("gax.Sleep: %v", err)
		}
	}
	_ = resp // TODO: use resp
```

Reply:
The returns are much nicer but the Example doesn't return errors, so it doesn't work. That's why I went with the closure in a previous patch.

I've sufficiently convinced myself that the closure is the better way to do this. This nesting of continues and breaks is horrible. Will take a minute later today to revert back to that style unless y'all have objections?

Reply:
Ahhh. You want the example to show up in GoDoc... I understand... Yeah, a closure that can return will be better than the nested breaks and stuff.

Sorry for the churn.

Reply:
Yes - sorry, I should have been more clear in the CL description. Churn is good - we're testing ideas. :) And, this is a bit hard to get right. Better we churn here than users churn trying to use it.

Done - PTAL and LMK what you think.